clio-server/src/main/scala/org/broadinstitute/clio/server/service/IndexService.scala

```diff
@@ -67,4 +90,69 @@ abstract class IndexService[CI <: ClioIndex](
       elasticsearchIndex
     )
   }
+
+  def validateUpsert(
+    existingMetadata: Json,
+    newMetadata: clioIndex.MetadataType
+  ): Either[Exception, clioIndex.MetadataType] = {
+    val differences = JsonDiff.diff(existingMetadata, newMetadata, remember = true)
+    /* Types of differences:
+        Replace - the existing value is different from what we want to set. We ignore nulls. This will overwrite.
+        Remove - the existing value doesn't exist in the new meta data. This will not overwrite.
+        Add - the existing value doesn't exist. This will not overwrite.
+        Move - effectively an Add and Remove. This will not overwrite.
+        Copy - effectively an Add. This will not overwrite.
+     */
+    val replaceOperations: Seq[Replace] = differences.ops.flatMap {
+      case replace: Replace => Some(replace).filter(_.value != Json.Null)
+      case _                => None
+    }
+
+    if (replaceOperations.isEmpty) {
+      Right(newMetadata)
+    } else {
+      val diffs: String = replaceOperations.map { replaceOp =>
+        s"Field: ${replaceOp.path}, Old value: ${replaceOp.old.getOrElse("")}, New value: ${replaceOp.value}"
+      }.mkString("\n")
+      Left(
+        UpsertValidationException(
+          s"""Adding this document will overwrite the following existing metadata:
+             |$diffs. Use '$forceString=true' to overwrite the existing data.""".stripMargin
+        )
+      )
+    }
+  }
+
+  def queryMetadataForKey(
+    indexKey: clioIndex.KeyType
+  ): Source[Json, NotUsed] = {
+    val keyJson = indexKey.asJsonObject
+    val fieldsToDrop = keyJson.keys.toSeq
+
+    searchService
+      .simpleStringQueryMetadata(indexKey.asJson)(
+        elasticsearchIndex,
+        queryConverter
+      )
+      .fold[Either[Exception, Option[Json]]](Right(None)) { (acc, storedDocs) =>
+        acc.flatMap {
+          case None =>
+            Right(Some(storedDocs.mapObject(_.filterKeys(!fieldsToDrop.contains(_)))))
+          case Some(_) =>
+            Left(
+              UpsertValidationException(
+                s"""Got > 1 ${clioIndex.name}s from Clio for key:
+                 |${indexKey.asJson
+                     .pretty(ModelAutoDerivation.defaultPrinter)}""".stripMargin
```

Comment:
Would pull this out of the string interpolation and into a local val, to avoid the weird line-break.